salt/utils/extend.py

```diff
@@ -0,0 +1,282 @@
+            with open(d, 'w') as dest_file:
+                dest_file.write(dest_contents)
+
+
+def _prompt_user_variable(var_name, default_value):
+    """
+    Prompt the user to enter the value of a variable
+
+    :param var_name: The question to ask the user
+    :type  var_name: ``str``
+
+    :param default_value: The default value
+    :type  default_value: ``str``
+
+    :rtype: ``str``
+    :returns: the value from the user
+    """
+    return click.prompt(var_name, default=default_value)
+
+
+def _prompt_choice(var_name, options):
+    """
+    Prompt the user to choose between a list of options, index each one by adding an enumerator
+    based on https://github.com/audreyr/cookiecutter/blob/master/cookiecutter/prompt.py#L51
+
+    :param var_name: The question to ask the user
+    :type  var_name: ``str``
+
+    :param options: A list of options
+    :type  options: ``list`` of ``tupple``
+
+    :rtype: ``tuple``
+    :returns: The selected user
+    """
+    choice_map = OrderedDict(
+        (u'{}'.format(i), value) for i, value in enumerate(options, 1)
+    )
+    choices = choice_map.keys()
+    default = u'1'
+
+    choice_lines = [u'{} - {} - {}'.format(c[0], c[1][0], c[1][1]) for c in choice_map.items()]
+    prompt = u'\n'.join((
+        u'Select {}:'.format(var_name),
+        u'\n'.join(choice_lines),
+        u'Choose from {}'.format(u', '.join(choices))
+    ))
+
+    user_choice = click.prompt(
+        prompt, type=click.Choice(choices), default=default
+    )
+    return choice_map[user_choice]
+
+
+def apply_template(template_dir, output_dir, context):
+    """
+    Apply the template from the template directory to the output
+    using the supplied context dict.
+
+    :param src: The source path
+    :type  src: ``str``
+
+    :param dst: The destination path
+    :type  dst: ``str``
+
+    :param context: The dictionary to inject into the Jinja template as context
+    :type  context: ``dict``
+    """
+    _mergetreejinja(template_dir, output_dir, context)
+
+
+def run(extension=None, name=None, description=None, salt_dir=None, merge=False, temp_dir=None):
+    """
+    A template factory for extending the salt ecosystem
+
+    :param extension: The extension type, e.g. 'module', 'state', if omitted, user will be prompted
+    :type  extension: ``str``
+
+    :param name: Python-friendly name for the module, if omitted, user will be prompted
+    :type  name: ``str``
+
+    :param description: A description of the extension, if omitted, user will be prompted
+    :type  description: ``str``
+
+    :param salt_dir: The targeted Salt source directory
+    :type  salt_dir: ``str``
+
+    :param merge: Merge with salt directory, `False` to keep seperate, `True` to merge trees.
+    :type  merge: ``bool``
+
+    :param temp_dir: The directory for generated code, if omitted, system temp will be used
+    :type  temp_dir: ``str``
+    """
+    assert HAS_CLICK, "click is not installed, please install using pip"
```

Comment:
Won't this bubble an `AssertionError` and respective traceback up to the users's console?
Maybe print the error and `exit(1)`?